### ISSUE TITLE
Batch segment reduce function

### ISSUE THREAD

**Issue by zegzag** [tensorflow/tensorflow#49043]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.15 (It's natural to update the feature in tf 2.x. And it's much better if the feature can be added in tf 1.x too)
- Are you willing to contribute it (Yes/No): Yes. I would try.

**Describe the feature and the current behavior/state.**
   Thanks for providing tensorflow framework and it's quit efficient to use `tf.math.segment_x` for some manipulations. Howerver,  we usually use batch training in deep learning pipeline. So, I'm eager for the function family like `tf.math.batch_segment_x`. Here is the api description for one example:
Batch segment reduce sum:
`tf.math.batch_segment_sum(data, batch_segment_ids, max_seq_len, mask_vals, name)`
- `data`: Tensor whose rank >= 2  
- `batch_segment_ids`: Tensor or list whose rank == 2; Sorted; `batch_segment_ids.shape[0] == data.shape[0]`
- `max_seq_len`: Scalar. Different instances in a batch may have different segmentation. Some length is larger(max value in `batch_segment_ids[i, :]`, some is small. This is for length alignment in a batch.  Clip while length of one instance(`data[i, :, ..., :]`) after op is larger than `max_seq_len`; Mask while smaller.
- `mask_vals`: Scalar or 1D tensor. If it's 1D tensor, it's length must be equal with `data.shape[0]`. Mask vals for instances in a batch. Scalar means all mask value is same. Tensor for different vals with different instances.

**Will this change the current api? How?**
Yes. Some new api will be added into `tf.math`, like `tf.math.batch_segment_sum`, `tf.math.batch_segment_max`, `tf.math.batch_segment_min`, etc.

**Who will benefit with this feature?**
Someone who often work with sequence model and need some feature engineering or some  complex manipulation in a model. 
Hope this feature request issue can be achieved and I appreciate a lot for all the development work about this issue.

**Any Other info.**
It would be better if `tf.math.batch_unsorted_segment_x` api can be added.

**Comment by ymodak** [comment]
Have you tried using (tensorflow bucketize op)[https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/bucketize]?
```python
from tensorflow.python.ops import math_ops
input = [[-5, 10000], [150, 10], [5, 100]]
boundaries = [0, 10, 100]
new_tensor = math_ops._bucketize(input, boundaries=boundaries)
print(new_tensor)
==>
tf.Tensor(
[[0 3]
 [3 2]
 [1 3]], shape=(3, 2), dtype=int32)
```

**Comment by zegzag** [comment]
Well, I learned that `math_ops._bucketize` is for catergorize the batch_data. This op doesn't change the shape of the input tensor. But `math.batch_segment_x`(or `match.batch_unsorted_segment_x`) will change each vector size along the first dimension of the input tensor(batch_size dimension) and do the size alignment. `math.batch_segment_x` is the extension of `math.segment_x` for batch data. 

I've tried to figure out using `math_ops._bucketize` to solve the case but failed.

Actually, I've found the _solution_ using the combination of `tf.while_loop` and `math.segment_x` for the case, loop along the first dimension, do `math.segment_x` for each vector and do size alignment. But this is time-consuming in _forward stage_ during training and prediction, because it will loop along the batch dimension. Thus, I'm eager for the low-level api for the case that can fully utilize fast characteristic of C++ to avoid time-consuming issue.

Here is the _solution_ code, hope this can help you for logic understanding and development work:
```
def batch_segment_op(batch_data, 
                     batch_segment_ids,
                     align_size,
                     min_data_val = None,
                     max_data_val = None,
                     reduce_type='sum', name=None, **kwargs):
    """batch segment reduce operation for seq tensor. 
    
    This is the extention of `tf.math.segment_x` in batch scene. Segment reduction along the SECOND dimension(seq dimension) in batch.
    Args:
      batch_data: Tensor; rank>=2; Shape: (N, S0,...Sn)
      batch_segment_ids: Tensor; rank==2; Shape: (N, S)
      align_size: Scalar; max_len after segment reduce for each instance in a batch
      reduce_type: str; x in {'sum','max','min','mean','prod'} for tf.math.segment_x
    """
    segment_op = None
    if reduce_type == 'sum':
        segment_op = tf.math.unsorted_segment_sum
    elif reduce_type == 'max':
        if min_data_val is None:
            raise ValueError("""You have to specify `min_data_val` when reduce
                              type is 'max' for masking default value(which is numeric_limits<T>::lowest())
                              of missing seg_ids to zero""")
        segment_op = tf.math.unsorted_segment_max
    elif reduce_type == 'min':
        if max_data_val is None:
            raise ValueError("""You have to specify `max_data_val` when reduce
                              type is 'min' for masking default value(which is numeric_limits<T>::max())
                              of missing seg_ids to zero""")
        segment_op = tf.math.unsorted_segment_min
    elif reduce_type == 'mean':
        segment_op = tf.math.unsorted_segment_mean
    elif reduce_type == 'prod':
        segment_op = tf.math.unsorted_segment_prod
    else:
        raise ValueError("Unsupported type for `reduce_type`; Only support in {'sum','max','min','mean','prod'}")
    
    batch_segid_min = tf.reduce_min(batch_segment_ids, axis=1, keepdims=True)
    batch_segment_ids = batch_segment_ids - batch_segid_min
    batch_num_segments = tf.reduce_max(batch_segment_ids, axis=1)
    batch_num_segments = tf.math.maximum(batch_num_segments, align_size)+1
    
    output_ta = tf.TensorArray(dtype=batch_data.dtype, 
                               size = 0,
                               dynamic_size = True)
    def cond(batch_data, 
             batch_seg,
             align_size, output, i):
        return tf.less(i, tf.shape(batch_data)[0])
    
    def body(batch_data, 
             batch_seg,
             align_size, output, i):
        seg_data = segment_op(batch_data[i], batch_segment_ids[i], batch_num_segments[i])
        seg_data = seg_data[:align_size]
        if reduce_type == 'min':
            seg_data = tf.where(seg_data <= max_data_val, 
                                seg_data, 
                                tf.zeros_like(seg_data))
        if reduce_type == 'max':
            seg_data = tf.where(seg_data >= min_data_val,
                                seg_data,
                                tf.zeros_like(seg_data))
        output = output.write(i, seg_data)
        return batch_data, batch_seg, align_size, output, i+1

    _, _, _, output_op, _ = \
        tf.while_loop(cond, body, 
                      [batch_data, 
                       batch_segment_ids,
                       align_size, output_ta, 0]
                     )
    return output_op.stack(name=name)
```

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
